readme.md
---

---

# Recordcleaner and Filer

This pipeline consists (at the moment) of two independent but complementary steps:
* RECORDCLEANER:  takes raw occurrence records and cleans them to standardised format. The main feature here is that we merge duplicate collections, while retaining all the information of the different iso-specimens.
* RECORD-FILER: This step integrates the data cleaned in RECORDCLEANER into a specified database (might at some point be automatically into a postgres/SQL database).

## Installation

You need the following prerequisites for using these scripts: python3 and conda. To install the required python-dependencies, 
in your terminal window navigate to your folder with the enclosed *'environment.yml'* file. Once there, execute the command *conda env create -f environment.yml*.


## Standardising occurrence records: RECORDCLEANER

<!-- ![There would be a funny picture here normally](TMP_titleimage.png "") -->


To launch recordcleaner, you can use the useful bash launcher, in terminal in your directory: (if you haven't already, make the
bash file executable -> uncomment first line below)

```
# chmod +x launch.sh
./launch.sh
```

This handy script let's you specify your input and output options, as well as other important parameters.
Alternatively, the script can be called directly in the command line with the '-h' flag, which lets a user know exactly which parameters are required and which are optional.

```
> python ./3_scripts/recordcleaner.py -h

RECORDCLEANER takes your raw occurence records and compares them to an existing database if you have it, cleaning up column names, removing duplicates and making it more pleasing in general

positional arguments:
  input_file            Raw input file path
  {GBIF,P}              File format. So far I can only handle Darwin core (GBIF) or herbonautes (P)
  {EXP, NO}             Input file is expert file or not (only arguments {EXP, NO} are accepted)
  working_directory     the directory in which to deposit all intermediate working files. Files that need reviewing start with "TO_CHECK_"
  output_directory      the wished output directory
  prefix                prefix for ouput filenames

optional arguments:
  -h, --help            show this help message and exit
--nonamecln NONAMECLN
                       if specified, collector names are expected in the format <Surname>, <F>irstname, or if deviating standardised to be identical across all datasets.

If it doesn't work, or you would like to chat, feel free to contact serafin.streiff<at>ird.fr
```

## Working with RECORDCLEANER

### Input files

In it's present form, the pipeline only takes input file in either the "darwin core format", as for example found in GBIF data, or
the format used at P for their herbonautes metadata collection projects, the BRAHMS database format used at Naturalis or data
extracted from Tropicos. Specially reduced expert data can be integrated using the 'SMALLEXP' flag in the expert_file parameter.
GBIF raw data downloads are in tab-separated tables (csv), and herbonautes data is separated by semicolons ';'. 

### Working directory

The working directory is specified, as some data might be non-conforming to the standardisation steps, and therefore not filterable with some steps of the pipeline. However, a lot of this might be still useable. Therefore it is written into the working directory for manual editing of the steps that cannot be done automatically.

For example, when standardising collector names (which is crucial for detecting duplicates in subsequent steps), I cannot handle names that are not in the format of some firstname, any middle names and some surname. E.g. if collections are filed under a program name (i.e. in SE-Asia, herbarium specimens are frequently labelled and numbered as something similar to "Forest Research Institute (FRI)", which I haven't so far been able to cleanly filter). Therefore it is faster to manually cross check these for consistency within dataset, and then I  reinsert them with no problem).
During the name standardisation step I output the concerned records to a temporary file, and after this step I plan to let the
user know than one can edit the records before reinserting them and continuing. 

During deduplication, suspected duplicates with large variation between coordinates are written to a file for manual inspection,
and not kept in the pipeline.


## What does record cleaner really do?

RECORDCLEANER goes through a few iterative step, which I briefly expain here.

* Step A:
  * A1: Standardise column names, remove unwanted columns and add empty columns we need later
    Note that as the postgreSQL database columns make all capital letters to small, I have changed this accordingly in the preprocessing
  * A2: Standardise data within some columns, e.g. separate all dates into separate columns day, month and year, make sure all barcodes have the institution leading before the number, have the first collector in a separate column,
  * A3: Standardise collector names to  *Streiff, SJR*, instead of *Serafin J. R. Streiff* in one record and *Streiff, Serafin* in another record
  * A4:  Standardise collector names even more by querying the Harvard Univ. Herbarium botanists
    [https://kiki.huh.harvard.edu/databases/botanist_index.html] database to get full or normalised names with a link to the
    record in that database (and wikipedia links for very famous botanists...). Names that are not found in that database are
    returned in the same format as the regex names (i.e. *Streiff, SJR*), whereas successfully found names are either returned as
    full names (*Surname, Firstname Any Middlename*) or abbreviated names (*Surname, F. A. M.*). The HUH database query is
    performed only with the surname, and results filtered with the original label data provided by the input data. HUH also
    returns, if available, information on collection regions of a botanist and wiki links.

* Step B:
  * B1: run some statistics on duplicates in the dataset. Select and remove all records where the collection number is a variation of <<s.n.>>. These are then treated separately to the others with other criteria to find duplicated specimens.
  * B2: remove said duplicates. The duplicates are at the moment identified by the following criteria: If the collector name (mix of HUH where possible, and regex where not), number, number sufix and year are identical, they are flagged as duplicates.
  Note that records with no collection number (i.e. *s.n.*) are treated separately (B3) 
  * B3: remove duplicates from separated *s.n.* -records. Here more other values are taken into consideration,  the combination of Surname, Collection Year, Month and Day and Genus + specific_epithet are used to identify duplicates. This leads to errors, but in my humble opinion it's better than nothing.
  * B4: Crossfill country full name and ISO codes (needed later on, and not always complete.)

* Step C:
  * Check taxonomy for accurracy, and update if necessary. At the moment this is done by cross checking with POWO (powo.kew.org), which for Annonaceae we can update relatively easily within the framework of our project collaborators. With other families the situation might be different, but changes can always be pushed by making the curators of POWO aware of published taxonomic changes that aren't up to date.
  * Check coordinates. Probably we just check for country centroids and points in the water. This will be done with already available packages, and issues flagged for potential correction in e.g. QGIS (qgis.org)
  This process is invoked as a separate step in R, as the packages available there are more used and robust (and I am more familiar with them). For the moment we implemented an automatic CoordinateCleaner (https://ropensci.github.io/CoordinateCleaner/index.html)

The resulting data of these cleaning steps are hte following files:
* FILENAME_cleaned.csv: final output from python processing (up to, including Step C1)
* FILENAME_spatialvalid.csv: final output with spatialvalid tag from *CoordinateCleaner*. Final data for import into database.

RECORD-FILER then goes and takes freshly (or even old) cleaned data and tries to integrate it into a pre-existing database

* RECORD-FILER:
  * Merge newly cleaned data with the database. Before the actual merging, I check for duplicates and merge these 
  * First all databases are read. These include two backlog datasets, the *indet*-backlog and the *no-coordinate*-backlog, as well as the master database. Backups of these are saved with a timestamp, so changes can be reverted.
  * With barcode priority, duplicates are checked and merged identically to the cleaning steps outlined above. If duplicates are found to with new data and the backlogs, these records are integrated into the new master version.
  * Finally the newly modified backlogs and master database are saved.




## Options

### Expert flag

CAUTION: this setting gives new data more priority than previous data. Therefore, make sure this data is clean, validated and doesn't include missing data!

If the EXP=EXP in the config file, then the pipeline expects this dataset to be completely expert determined and curated. The values here are given precedence over duplicates found in the master database. The main utility of using this flag is to integrate this data to update any determinations and georeferencing.

if EXP=SMALLEXP, then records are found with the barcode, and any relevant data is overwritten with the new data. Make sure that
these datasets are therefore correct!

### SQL/postgres interface

A small script is provided which automatically up/downloads a master database from a postgres server (y_sql_functions.py).



## TODO

* Keep readme uptodate with new developments and changes
* **add overwrite protection in case script is called multiple times**, at least for time intensive steps (removed for debugging!!) --> done as mode='x' for example within pd.to_csv()
* **DONE**: RETURN interactive input(), removed for scripting...
* **DONE**: Barcodes error in GBIF input! (issue dealing with "-" within barcode.)
* **DONE** - implement problem name output and reinsertion, optionally pausing execution
* **DONE** when do we query POWO/IPNI?? irrelevant in my opinion. We do it before inserting a new bunch of data into the master database. 
* quantify fast and slow steps and make backup files between, so we can restart at that step (maybe integrate variable to call analysis from step XYZ)
* **DONE** make distinction between non-spatialvalid and no coordinate data!! then we can revise non-spatialvalid points in QGIS

* **Implement background files**:
  * **DONE**  indets and similar: before integrating data, check for data previosly set aside because we had no conclusive data...?

  * **DONE**  master distribution database for integration.


# Common problems / error sources

Some data sources have aspects that can lead to errors in the pipeline. Usually these can be found by checking the error message
at or near the end of the log file, or in the terminal.
Common error sources are for example the occurrence of ';' within a cell which is not in quotes, or quote characters that are
misplaced (e.g. Coordinates in text written as e.g. 22°46'11\"N 104°49'07\"E, where the escape character \" is not correctly
identified.
This leads to the message **Error tokenizing data.**, followed by a mention of the problematic line in the input file.  This can
be solved by manually editing the input.

Also, make sure not to run 2 separate instances of the record-filer step simultaneously, as this might create a chaos between the
master database files!

# Future updates and modifications possible

For advanced users, one might change the columns to subset, for example to integrate data from other sources. The columns are specified in the
file "./3_scripts/z_dependencies.py". **It is crucialthat both the columns and their associated datatype are correctly specified
there.** 
Failure to do so will create problems when
working with the data. Details and data-source specific modifications necessary to standardise data can be changed in the script *z_functions_a.py*

**more to come**


changelog.md
# changelog RECORDCLEANER


# 2023-08-01:
    - switched from iso2 to iso3 in country_id (works directly with coordinate_cleaner)...

# 2023-08-02:
    - changed the barcode formatting to include alphanumeric repeating formats of 00000A000A000  (to be validated)
    - changed indet_backlog criteria from 'status' to 'accepted_name': if NA, then in backlog.

# 2023-08-08:
    - added 'link' column, where urls leading to the image can be integrated (remains hit-and-miss for GBIF, as there data entered quite randomly sometimes)
    
# 2023-08-09:
    - added 'BRAHMS' flag, allows integration of BRAHMS extract data
        included slight modifications to name standardisation, recurring NA vs NA-string 

# 2023-08-22:
    - added basic scripts for SQL integration
    - finalised brahms integration

# 2023-08-23:
    - merged branch brahms integration
    - new branch: speed up master merging

# 2023-09-13:
    - new branch deduplication error solving (coordinates);
        * [solved] discovered bug in deduplication leading to data getting lost in groupby (groupby on a column including only NA make data go away)
        * [solved]bug in (pure) colnum not extracting correctly for the format 00-123 or similar...
        * [BUG] coordinate filter (difference between 0.1 mean, otherwise throw out, not working)

# 2023-09-18:
    - coordinate problems worked out. now coordinates more than 0.1 variance are tested for country-coordinate match and if match then kept...

# 2023-09-25:
    - [SOLVED] coordinates now filter eventual duplicates with coordinates with a variance > 0.1

# 2023-09-26/27:
    - cleaned up superfluous files, deprecated scripts...
    - merging branch coordinate debug into master

3_scripts/z_expert.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Functions for expert det

2023-08-23 sjrs


- format input data

- check for missing data etc.

- integrate with master/find duplicated BC
"""

import codecs
import pandas as pd
import numpy as np
import logging
import pykew.ipni as ipni
import swifter              # not used at the moment


from pykew.ipni_terms import Filters as ipni_filter
from pykew.ipni_terms import Name as ipni_name              # for taxonomic additional info
from tqdm import tqdm                                       # progress bar for loops

def read_expert(importfile, verbose=True):
    """
    read file, check columns
    """
    print('EXPERT file integration. \n',
          'Please assure that your columns are the following:',
          'ddlat, ddlong, locality, country or ISO2, recorded_by, colnum_full, det_by, det_date, barcode')
    imp = codecs.open(importfile,'r','utf-8')
    exp_dat = pd.read_csv(imp, sep = ';',  dtype = str)
    exp_dat['source_id'] = 'specialist'

    # make prefix from colnum
    exp_dat['prefix'] = exp_dat.colnum_full.str.extract('^([a-zA-Z]*)')
    exp_dat['prefix'] = exp_dat['prefix'].str.strip()

    # make sufix from colnum
    # going from most specific to most general regex, this list takes all together in the end
    regex_list_sufix = [
        r'(?:[a-zA-Z ]*)$', ## any charcter at the end
        r'(?:SR_\d{1,})', ## extract 'SR_' followed by 1 to 3 digits
        r'(?:R_\d{1,})', ## extract 'R_' followed by 1 to 3 digits
    ]
    exp_dat['sufix'] = exp_dat['colnum_full'].astype(str).str.extract('(' + '|'.join(regex_list_sufix) + ')')
    exp_dat['sufix'] = exp_dat['sufix'].str.strip()

    # extract only digits without associated stuff, but including some characters (colNam)
    regex_list_digits = [
        r'(?:\d+\-\d+\-\d+)', # of structure 00-00-00
        r'(?:\d+\s\d+\s\d+)', # 00 00 00 or so
        r'(?:\d+\.\d+)', # 00.00
        r'(?:\d+)', # 00000
    ]
    exp_dat['colnum']  = exp_dat.colnum_full.str.extract('(' + '|'.join(regex_list_digits) + ')')
    exp_dat['colnum'] = exp_dat['colnum'].str.strip()



# det date
    exp_dat[['det_year', 'det_month', 'det_day']] = exp_dat['det_date'].str.split("-", expand=True)



    exp_dat[['huh_name', 'geo_col', 'wiki_url']] = '0'
    exp_dat['orig_recby'] = exp_dat['recorded_by']
    exp_dat['col_year'] = pd.NA

    return exp_dat


# do HUH
# get ipni numbers?




# deduplication....
# we will do this by barcodes and or recorded_by + colnum + ??

def deduplicate_small_experts(master_db, exp_dat, verbose=True):
    """
    Find duplicates based on barcode, and collector name,

    Any values in these records found are overwritten by 'expert' data. This is assuned to be of (much) better quality.
    """

    # first some housekeeping: remove duplicated barcodes in input i.e. [barcode1, barcode2, barcode1] becomes [barcode1, barcode2]
    exp_dat.barcode = exp_dat.barcode.apply(lambda x: ', '.join(set(x.split(', '))))    # this combines all duplicated barcodes within a cell
    master_db.barcode = master_db.barcode.apply(lambda x: ', '.join(set(x.split(', '))))    # this combines all duplicated barcodes within a cell

    # drop any determinations that are empty. This would make a mess.
    exp_dat = exp_dat[pd.notna(exp_dat.accepted_name)] 
    # add column for flag if data found or not
    exp_dat['matched'] = '0'

    #----- prep barcodes (i.e. split multiple barcodes into seperate values <BC001, BC002> --> <BC001>, <BC002>)
    # split new record barcode fields (just to check if there are multiple barcodes there)
    bc_dupli_split = exp_dat['barcode'].str.split(',', expand = True) # split potential barcodes separated by ','
    bc_dupli_split.columns = [f'bc_{i}' for i in range(bc_dupli_split.shape[1])] # give the columns names..
    bc_dupli_split = bc_dupli_split.apply(lambda x: x.str.strip())
    # some information if there are issues
    logging.debug(f'NEW OCCS:\n {bc_dupli_split}')
    logging.debug(f'NEW OCCS:\n {type(bc_dupli_split)}')
    # repeat for master data
    # split potential barcodes separated by ','
    master_bc_split = master_db['barcode'].str.split(',', expand = True) 
    master_bc_split.columns = [f'bc_{i}' for i in range(master_bc_split.shape[1])]
    master_bc_split = master_bc_split.apply(lambda x: x.str.strip())  # strip all leading/trailing white spaces!
    logging.debug(f'master OCCS:\n {master_bc_split}')

    # to make an exceptions dataframe get structure from master
    exceptions = master_db.head(1)

    # for progress bar in console output
    total_iterations = len(exp_dat)
    print('Crosschecking barcodes...\n', total_iterations, 'records in total.')
    for i in tqdm(range(total_iterations), desc = 'Processing', unit= 'iteration'):
        # the tqdm does the progress bar

        barcode = list(bc_dupli_split.loc[i].astype(str))
            # logging.info(f'BARCODE1: {barcode}')
        # if multiple barcodes in the barcode field, iterate across them
        for x in  range(len(barcode)):
            bar = barcode[x]
            if bar == 'None':
                # this happens a lot. skip if this is the case.
                a = 'skip'
                #logging.info('Values <None> are skipped.')
            else:
                selection_frame = pd.DataFrame()  # df to hold resulting True/False mask  
                # now iterate over columns to find any matches
                for col in master_bc_split.columns:
                    # iterate through rows. the 'in' function doesn't work otherwise
                    #logging.info('checking master columns')
                    f1 = master_bc_split[col] == bar # get true/false column
                    selection_frame = pd.concat([selection_frame, f1], axis=1) # and merge with previos columns
                # end of loop over columns

                # when selection frame finished, get out the rows we need including master value
                sel_sum = selection_frame.sum(axis = 1)
                sel_sum = sel_sum >= 1 # any value >1 is a True == match of barcodes between dataframes 
            
                if sel_sum.sum() == 0:
                    # logging.info('NO MATCHES FOUND!')
                    out_barcode = pd.DataFrame([bar])

                # in this case we do not modify anything!

                else:
                    out_barcode = pd.Series(master_db.barcode[sel_sum]).astype(str)
                    out_barcode.reset_index(drop = True, inplace = True)


                # replace i-th element of the new barcodes with the matched complete range of barcodes from master
            
                input = str(exp_dat.at[i, 'barcode'])
                master = str(out_barcode[0])
                new = input + ', ' + master

                # reduce duplicated values
                new = ', '.join(set(new.split(', ')))

                # QUICK CHECK if recorded by and colnum are identical. Flag and save to special file if not
                print('1', master_db.loc[sel_sum, 'recorded_by'].item())
                print('2', exp_dat.loc[i, 'recorded_by'])
                if master_db.loc[sel_sum, 'recorded_by'].item() == (exp_dat.loc[i, 'recorded_by']):

                    #----------------------- Imperative Values. Missing is not allowed ----------------------#
                    if exp_dat.loc[i,['accepted_name', 'det_by', 'det_year']].isna().any() == True:
                        # make a visible error message and raise exception (abort)
                        print('\n#--> Something is WRONG here:\n',
                              '\n#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#\n',
                              exp_dat.loc[i,['accepted_name', 'det_by', 'det_year']],
                              '\n#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#\n',
                              '\n I am aborting the program. Please carefully check your input data.\n',
                              '\n#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#-#\n',)
                        raise(Exception('One of \'accepted_name\', \'det_by\', or \'det_year\' is NA.\n',
                                        'I do not allow such records as expert data....'))
                    # otherwise the crucial data is here so we can proceed...
                    master_db.loc[sel_sum, 'barcode'] = new
                    master_db.loc[sel_sum, 'accepted_name'] = exp_dat.at[i, 'accepted_name'] 
                    master_db.loc[sel_sum, 'det_by'] = exp_dat.at[i, 'det_by'] 
                    master_db.loc[sel_sum, 'det_year'] = exp_dat.at[i, 'det_year'] 
                    
                    #----------------------- Facultative Values. Missing is allowed --------------------------#
                    if ~np.isnan(exp_dat.loc[i, 'ddlat']):
                        master_db.loc[sel_sum, 'ddlat'] = exp_dat.at[i, 'ddlat'] 
                    if ~np.isnan(exp_dat.loc[i, 'ddlong']):
                        master_db.loc[sel_sum, 'ddlong'] = exp_dat.at[i, 'ddlong'] 

                    #----------------------- Automatic Values. Filled anyway --------------------------#
                    master_db.loc[sel_sum, 'status'] = 'ACCEPTED'
                    master_db.loc[sel_sum, 'expert_det'] = 'A_expert_det_file'
                    master_db.loc[sel_sum, 'prefix'] = exp_dat.at[i, 'prefix'] 

                    #----------------------- Cosmetic Values. Missing is allowed --------------------------#
                    # print((pd.Series(exp_dat.loc[i, 'colnum_full']).isin(['-9999','NA','<NA>','NaN'])).any())
                    # print('CNF', exp_dat.loc[i, 'colnum_full'])
                    if pd.Series(exp_dat.loc[i, 'colnum_full']).isna().any():
                        print('yes NA')
                        master_db.loc[sel_sum, 'colnum_full'] = exp_dat.at[i, 'colnum_full'] 
                    if pd.Series(exp_dat.loc[i, 'locality']).isna().any():
                        master_db.loc[sel_sum, 'locality'] = exp_dat.at[i, 'locality'] 

                    exp_dat.loc[i, 'matched'] = 'FILLED'


                else:
                    # exception where 'recorded_by' do not match
                    new_except = pd.concat([master_db.loc[sel_sum], pd.DataFrame(exp_dat.loc[i]).transpose()]) # see if this works...
                    # drop the offending rows for manual editing and adding later
                    master_db = master_db.loc[~sel_sum]
                    try:
                        # if exception already exists add concat
                        exceptions  = pd.concat([exceptions, new_except])
                    except:
                        # otherwise new 
                        exceptions = new_except
                # at the end of for loop print exceptions to file, let me manually redo it.
    print('MATCHED?', exp_dat.matched)
    exp_dat_to_integrate = exp_dat[exp_dat.matched != 'FILLED']
    print(exp_dat_to_integrate)

    # as we may have some data remaining, we just append it ot the master, as it's expert perfect ( in theory for our purposes)
    master_db =  pd.concat([master_db, exp_dat_to_integrate])
    ## TODO## TODO## TODO## TODO## TODO## TODO## TODO## TODO## TODO## TODO

    return master_db, exceptions


def integrate_exp_exceptions(integration_file, exp_dat):
    """
    read and concatenate data manually edited, chekc data lengths 
    """
    imp = codecs.open(integration_file,'r','utf-8') #open for reading with "universal" type set
    re_exp = pd.read_csv(imp, sep = ';',  dtype = str, index_col=0) # read the data
    try:
        new_exp_dat = pd.concat([re_exp, exp_dat])
        if len(new_exp_dat) == (len(re_exp) + len(exp_dat)):
            print('Reintegration successful.')
            logging.info('reintroduction successful')
        else:
            print('reintegration not successful.')
            logging.info('reintegration unsuccessful')
    
    except:
        new_exp_dat = exp_dat
        print('reintegration not successful.')
        logging.info('reintegration unsuccessful')

    return new_exp_dat




def expert_ipni(species_name):
    ''' 
    Check species names against IPNI to get publication year, author name and IPNI link

    INPUT: 'genus'{string}, 'specific_epithet'{string} and 'distribution'{bool}
    OUTPUT: 'ipni_no' the IPNI number assigned to the input name
            'ipni_yr' the publication year of the species

    '''
    #print('Checking uptodate-ness of nomenclature in your dataset...')

    query = species_name
        #print(query)
    res = ipni.search(query, filters = ipni_filter.specific) # so we don't get a mess with infraspecific names
        #res = ipni.search(query, filters=Filters.species)  # , filters = [Filters.accepted])
    try:
        for r in res:
            #print(r)
            if 'name' in r:
                r['name']
        ipni_pubYr = r['publicationYear']
        ipni_no = 'https://ipni.org/n/' + r['url']
        ipni_author = r['authors']
        #print(ipni_pubYr)
        #logging.debug('IPNI publication year found.')
    except:
        ipni_pubYr = pd.NA

    return ipni_pubYr, ipni_no, ipni_author




def exp_run_ipni(exp_data):
        """
        wrapper for swifter apply of above function 'expert_ipni()'
        """
        print(exp_data.columns)
        exp_data[['ipni_species_author', 'ipni_no', 'ipni_pub']] = exp_data.swifter.apply(lambda row: expert_ipni(row['accepted_name']), axis = 1, result_type='expand')

        return exp_data






# # # # # # # --- DEBUGGING LINES ----- # # # # # # # 

# # debug_master = pd.read_csv('/Users/serafin/Sync/1_Annonaceae/G_GLOBAL_distr_DB/X_GLOBAL/debug/smallexp_debug.csv', sep =';')
# debug_exp_file = '/Users/serafin/Sync/1_Annonaceae/G_GLOBAL_distr_DB/X_GLOBAL/debug/exp.csv'

# # print(debug_exp_file)
# # print(debug_master)
# # print('---WORKING---')
# debug_exp = read_expert(debug_exp_file)
# print(debug_exp)

# # print(debug_exp)
# debug_exp = exp_run_ipni(debug_exp)

# print(debug_exp)
# # final, exception = deduplicate_experts_minimal(debug_master, debug_exp)

# # print(final.accepted_name)
# # print('EXCEPTIONS', exception)
# spec = 'Cananga odorata'
# test_Y, test_N, testA = expert_ipni(spec)
    # print(test_Y, test_N, testA)